### PR TITLE
Make AST serializable

### PR DESCRIPTION
**Pull request by AlicanC**
- Implements `serde::Serialize` for `sway_ast::Module` and such.
- Adds a simple integration test to `sway_parse` which uses `insta` for RON snapshot testing.

This should allow us to dump AST in more convenient formats and also use tools like `insta` to do testing on the serialized output.

Note: Deserialization was left for later.

### CODE REVIEW

**Comment by AlicanC** [comment]
@kayagokalp will it be necessary to keep `no-span-debug` after this? If not I can remove that here.

**Comment by kayagokalp** [comment]
> will it be necessary to keep `no-span-debug` after this? If not I can remove that here.

I think the use case for the `--no-span-debug` still exists: I originally needed it while printing out entire namespaces for debugging and spans were taking %90 of the output. So I would say we should keep it but good with it either way.

**Comment by AlicanC** [comment]
> Is the idea to add more AST tests in this style? And what is the idea around adding deserialization?
> 
> 
> 
> Overall this seems fine, but I'd be wary about building more testing infrastructure on this as it exposes all the internal details of the AST. And since it's all opaque to the compiler, every little refactoring will cause these tests to fail, making them a bit brittle.

Not planning to add any tests in this E2E style (maybe one for each module type?), but I've found some unit tests for parsing complicated stuff like PathExpr and UseItem necessary.

I'm aware how brittle this is, but: is this okay because we eventually want a clean/stable AST serialization and this will push us towards that? Do I want to mess freely with the AST or do I want that test to keep me in check so when I'm working on AST/parsing I'm forced to be conscious on how it's serialized?

On deserialization: I had to just discard some info during serialization to avoid refactoring too much so the format is lossy. Refactoring Span and such should open the way for it.

File: sway-ast/Cargo.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ description = "Sway's parser"
 
 [dependencies]
 extension-trait = "1.0.1"
-num-bigint = "0.4.3"
+num-bigint = { version = "0.4.3", features = ["serde"] }
 num-traits = "0.2.14"
+serde = { version = "1.0", features = ["derive"] }
```
Comment (kayagokalp) [inline sway-ast/Cargo.toml:15]: Hmm I wonder if we should have a `serde` feature that enables this just like most of the crates we depend on. The serde feature would enable the actual serialization implementations. 

I am imagining serialization is not needed by the compiler itself but it is a good option to offer to downstream crates (for example lsp).

**Review by kayagokalp — COMMENTED** [review]
Looking good! I am not sure where to stand in terms of consuming this serialization in tests as it would be hard to maintain. At the same time it seems like a good idea to somehow enforce AST stabilization as it could lead us to be aware of breaking changes from library perspective (although we are not doing that right now, lately I started dreaming about [this](https://github.com/FuelLabs/sway/issues/1950) 😄) 

That being said I think this is an improvement even we end up not using the tests as downstream crates may want to (de*)serialize the AST and if we let them have this option with a feature, I feel like it would be a good practice overall

File: sway-ast/Cargo.toml (hunk 1 above)
Comment (AlicanC) [reply sway-ast/Cargo.toml:15]: Compiler could have a parse cache, no?

Comment (kayagokalp) [reply sway-ast/Cargo.toml:15]: Hmm that makes sense 👌

File: sway-types/src/ident.rs [hunk 2]
```diff
@@ -81,6 +83,18 @@ impl BaseIdent {
 /// often be different.
 pub type Ident = BaseIdent;
 
+impl Serialize for Ident {
+    // Serialize an `Ident` struct with two fields: `to_string` and `span`.
```
Comment (kayagokalp) [inline sway-types/src/ident.rs:87]: Maybe these were intended to be doc comments? 
```suggestion
    /// Serialize an `Ident` struct with two fields: `to_string` and `span`.
```

File: sway-types/src/span.rs [hunk 3]
```diff
@@ -76,6 +78,18 @@ pub struct Span {
     path: Option<Arc<PathBuf>>,
 }
 
+impl Serialize for Span {
+    // Serialize a tuple two fields: `start` and `end`.
```
Comment (kayagokalp) [inline sway-types/src/span.rs:82]: ^
```suggestion
    /// Serialize a tuple two fields: `start` and `end`.
```

**Review by kayagokalp — COMMENTED** [review]
lgtm, left 2 non-important nits deferring to compiler team for the approvals

**Comment by IGI-111** [comment]
>I'm aware how brittle this is, but: is this okay because we eventually want a clean/stable AST serialization and this will push us towards that? Do I want to mess freely with the AST or do I want that test to keep me in check so when I'm working on AST/parsing I'm forced to be conscious on how it's serialized?

I'm not sure we want the AST to be stable to the point of serializing identically across versions. That seems like a lot more coupling than necessary to me.

More likely we want the structure to remain the same but the node content will very likely have to evolve.

**Comment by AlicanC** [comment]
I think we need to start keeping our AST stable after some point. There will be tools like [jscodeshift](https://github.com/facebook/jscodeshift) that work on ASTs that would break. Even today, a simple change in the parser will have you refactor all the way down to the formatter.

**Review by IGI-111 — APPROVED** [review]

**Review by mohammadfawaz — APPROVED** [review]

**Pull request merged**